### DISCUSSION TITLE
Custom bounding mesh

### DISCUSSION THREAD

**Discussion by GuangyanCai** [mitsuba-renderer/mitsuba3#1118] [Q&A] [answered]
Hi there, what's the best way to implement a custom bounding mesh in mitsuba? That is, I want to be able to do ray intersection with this bounding mesh while excluding it from the scene to be rendered. So it's going to be similar to `BoundingSphere3f` but with user defined triangle faces. I want to use it in a Python plugin. 

It looks like I can easily implement it with `ShapeKDTree` but I couldn't find it in the Python binding, which seems to be exported in `main_v.cpp`. How can I access it? Or do I have to go back to C++ to use it? 

Thanks in advance!

**Comment by njroussel** [comment, marked as answer]
Hi @GuangyanCai 

A few things: 

* The KDTree is not compiled by default, and when it is compiled it is only available in `scalar` variants.
* Typically you're using Embree (LLVM or scalar variants) or OptiX (in CUDA variants). Both frameworks typically use their own bounding box shape: AABBs. Note that these are just to help build the acceleration structure. I don't think that this is what you want. Unless you're experimenting with acceleration structures.
* We do not allow custom shape plugins. If you wish to write you own custom shape, you must do it in C++ (and CUDA for OptiX).

With these things being said, I think the easiest would be to just add your "Mesh bounding box"-shape to the scene with a `null` BSDF. 
By doing this, it will not be visible in the final render and you would still get intersections with that shape. I' don't know you want to do once you get those intersections, so I can't really tell what the next steps would be. In your integrator, you are somewhat free to do whatever you want from the `SurfaceInteraction3f` objects you would get back from the `scene.ray_intersect()` method.

**Reply by GuangyanCai** [reply]
Thank you! That answered my question :)
